multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,340 @@
+		return ctrl.Result{}, err
+	}
+	nsLabels, podLabels := ns.Labels, pod.Labels
+	return ctrl.Result{}, r.onPodUpdate(ctx, req.NamespacedName.String(), getNormalizedLabel(nsLabels, podLabels))
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Ignore status update event via GenerationChangedPredicate
+	instance := predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(instance).
+		Watches(&source.Kind{Type: &v1.Namespace{}},
+			handler.EnqueueRequestsFromMapFunc(namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc enqueues Namespace events to the reconciler.
+func namespaceMapFunc(o client.Object) []reconcile.Request {
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Name:      o.GetName(),
+				Namespace: o.GetNamespace(),
+			},
+		},
+	}
+}
+
+// onPodDelete removes the Pod and label identity relation from the cache, and updates the cluster's
+// label identity ResourceExport if necessary (the Pod deletion event causes some label identities
+// to no longer being present in the cluster).
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podKey, normalizedLabel string) error {
+	if resExportNeedUpdate := r.deleteLabelMapping(podKey, normalizedLabel); resExportNeedUpdate {
+		labelsInCluster := r.getLabelsInCluster()
+		if err := r.refreshLabelIdentityResourceExport(ctx, labelsInCluster); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *LabelIdentityReconciler) deleteLabelMapping(podKey, normalizedLabel string) bool {
+	r.labelMutex.Lock()
+	defer r.labelMutex.Unlock()
+
+	originalLabel, ok := r.podLabelCache[podKey]
+	if !ok {
+		// Might be a host-network Pod deletion
+		return false
+	} else if normalizedLabel == "" {
+		// Use the cached label identity for the Pod to clean up labelToPodsCache
+		normalizedLabel = originalLabel
+	}
+	delete(r.podLabelCache, podKey)
+	podNames, ok := r.labelToPodsCache[normalizedLabel]
+	if !ok || !podNames.Has(podKey) {
+		klog.V(2).InfoS("Label is already removed from cache", "Pod", podKey, "label", normalizedLabel)
+		return false
+	}
+	podNames.Delete(podKey)
+	if len(podNames) == 0 {
+		klog.V(2).InfoS("Label no longer exist in the cluster", "label", normalizedLabel)
+		delete(r.labelToPodsCache, normalizedLabel)
+		return true
+	}
+	return false
+}
+
+func (r *LabelIdentityReconciler) checkPodLabelCache(pod, normalizedLabel string) (string, bool) {
+	r.labelMutex.RLock()
+	defer r.labelMutex.RUnlock()
+
+	cachedPodLabel, labelChanged := "", false
+	cachedPodLabel, ok := r.podLabelCache[pod]
+	if !ok {
+		labelChanged = true
+	} else if cachedPodLabel != normalizedLabel {
```

Comment:
looks like this `else` can be removed as well.

Reply:
I think I would keep this check just to be safe